### ISSUE TITLE
Update Pymoo Interface to v0.6.0

### ISSUE THREAD

**Issue by jharris1679** [fau-advanced-separations/CADET-Process#10]
I have installed CADET-Process==0.6.3 via pip on python 3.7 in a docker container running Red Hat 4.8.3-9. I have CADET==0.10 installed, and pymoo==0.6.0. When I import CADETProcess, it fails when trying to import a submodule of pymoo.

>bash-4.2# python3
Python 3.7.13 (default, Jul 28 2022, 09:25:00)
[GCC 4.8.3 20140911 (Red Hat 4.8.3-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
$ import cadet
$ cadet.\_\_version\_\_
    '0.10'
$ import pymoo
$ pymoo.\_\_version\_\_
    '0.6.0'
$ import CADETProcess
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/var/task/CADETProcess/__init__.py", line 31, in <module>
    from . import optimization
  File "/var/task/CADETProcess/optimization/__init__.py", line 5, in <module>
    from .pymooAdapter import NSGA2, U_NSGA3
  File "/var/task/CADETProcess/optimization/pymooAdapter.py", line 11, in <module>
    from pymoo.util.termination.default import MultiObjectiveDefaultTermination
ModuleNotFoundError: No module named 'pymoo.util.termination'

**Comment by schmoelder** [comment]
Hi Josh, thanks for reporting this issue. The latest version of Pymoo included some breaking changes in the interface ([see here](https://pymoo.org/versions.html#version-0-6-0)). I will try to fix this in the next release. Until then, please downgrade Pymoo to v0.5.0

**Comment by schmoelder** [comment]
Hey,

good idea, pymoo's API does not seem too stable yet. 

I actually already upgraded the API yesterday on the dev branch. I'm currently in the process in testing CADET-Process 0.7.0 and hope to release it in the next couple of days. It includes a lot of improvements for the definition and solution, and post-processing of OptimizationProblems. 

If you're interested in testing, I'm happy to give some introduction. Feedback is also always welcome. :)

**Comment by jharris1679** [comment]
@schmoelder Just out of curiosity which version of python are you running for your testing? We are running 3.9, but I've had issues getting CP to run with 3.9, or 3.8. Looks like the hopsy dependency is the limiting factor, at least as far as I can tell, so I'm running it in 3.7. I see you've also upgrade to hopsy 1.0.0 in the dev branch, hopefully with that it'll run on 3.9.

**Comment by schmoelder** [comment]
I'm mostly running Python 3.9 on Ubuntu 20/22.04. But i took the opportunity to also add Windows, and Python 3.8 and 3.10 and so far they seem to be running fine. Can you elaborate what is the issue?

Hopsy - which btw is also developed at our institute - also recently had a new release (v.1.0.0) which included a new (stable) API. Again this has been already fixed and pinned in the dev branch. Unfortunately, they do still not provide MacOS binaries but they are aware of it. Hopefully it won't take too long for them to add that.

**Comment by jharris1679** [comment]
Thanks for the info, I'm running in an Amazon distribution of Red Hat because we are building a containerized application deployed in their Lambda functions. I was seeing that pip3 couldn't find a distribution for hopsy when running 3.9 or 3.8, but it could when running 3.7. It appears now though that may have been caused by another issue, I'm not seeing that behaviour anymore so moving forward on 3.9.

**Comment by schmoelder** [comment]
Ok, sounds interesting. Would you mind sharing some more detail what you're working on? We can also have a call some time soon.
Btw, do you know our [CADET-Forum](https://forum.cadet-web.de)? It's also a great place to exchange ideas.

**Comment by jharris1679** [comment]
Hey, yeah we are building a platform for simulating various processes in pharma manufacturing, chromatography being one of them. I did find the CADET forum, and currently I'm stuck installing SuiteSparse, so perhaps I'll reach out there for some help.

I'd be more than happy to have a call, very curious to learn more about CADET and the other projects you're working on, feel free to send me an invite at joshua@basetwo.ai

**Comment by mrsonne** [comment]
I just ran into the same problem with pymoo 0.6 & CADETProcess 0.6.3. What puzzles me is that it works fine on the Jupyter instance provided for the (educational and fun) workshop in October. Just for my understanding: were there some last minute local change or am I doing something wrong? BTW it works fine on my PC with pymoo 0.5.  Later, we could consider only importing e.g. pymoo when performing optimization. Cheers, Jacob

------
**From my PC**
 ```
ModuleNotFoundError: No module named 'pymoo.util.termination'
```
with 

```
CADET-Process.__version__ = '0.6.3'
```

and 

```
$ head -20 /mnt/c/Users/xxx/AppData/Roaming/Python/Python39/site-packages/CADETProcess/optimization/pymooAdapter.py
import os
import random
import time

import dill
import numpy as np

import pymoo
from pymoo.core.problem import Problem
from pymoo.factory import get_reference_directions
from pymoo.util.termination.default import MultiObjectiveDefaultTermination
```

**From the October workshop Jupyter instance**

```
jupyter-xxx@novonordisk.c-eb574@ubuntu18:~$ python3
Python 3.9.11 (main, Mar 28 2022, 10:10:35) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import CADETProcess
>>> print(CADETProcess.__file__)
/opt/tljh/user/lib/python3.9/site-packages/CADETProcess/__init__.py
>>> print("CADETProcess", CADETProcess.__version__)
CADETProcess 0.6.3
>>> import cadet
>>> print("cadet", cadet.__version__)
cadet 0.11
>>> import pymoo
>>> print("pymoo", pymoo.__version__)
pymoo 0.6.0
```

and 

```
jupyter-xxxx@novonordisk.c-eb574@ubuntu18:~$ head -20 /opt/tljh/user/lib/python3.9/site-packages/CADETProcess/optimization/pymooAdapter.py 
import importlib
import os
import random

import dill
import numpy as np

from pymoo.core.problem import Problem
from pymoo.util.ref_dirs import get_reference_directions
from pymoo.termination.default import DefaultMultiObjectiveTermination
from pymoo.core.repair import Repair

from CADETProcess.dataStructure import UnsignedInteger, UnsignedFloat
from CADETProcess.optimization import OptimizerBase


class PymooInterface(OptimizerBase):
    """Wrapper around pymoo."""
    supports_multi_objective = True
    supports_linear_constraints = True
```

**Comment by schmoelder** [comment]
Dear Jacob,

> I just ran into the same problem with pymoo 0.6 & CADETProcess 0.6.3.

Yes, they updated their interface and I should be a bit more diligent with specifying dependency versions.

> What puzzles me is that it works fine on the Jupyter instance provided for the (educational and fun) workshop in October. Just for my understanding: were there some last minute local change or am I doing something wrong?

First of all, thanks a lot! It was very fun, indeed! :)
On jupyter, I installed the `dev` branch which already has the updated interface.

I'm fully aware that things are not as clean as they should be here in the repo. In the last months so many things have changed under the hood that I never managed to merge everything from `dev` into `master`. Things are becoming more stable and now it's mostly (online) documentation that needs updating.
So if you want to test all the new features, feel free to use the `dev` branch. It also contains many other improvements. Be aware that I do sometimes `force-push` over there. But in future, we will strive to use more short-lived branches when adding new features. The current overhaul turned out much bigger than expected.
Thanks for understanding!

**Comment by mrsonne** [comment]
Thanks @schmoelder for the quick and detailed response. I fully understand and I'm exited that I finally have the time to start running CADET here.

**Comment by schmoelder** [comment]
closed in b7c17839f0cdfb12285e8a4dfd45f555f01a257b
